include/wil/registry_basic.h

```diff
@@ -77,1599 +77,1826 @@ namespace details
-}
+                if (return_value.inner_error == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND))
+                {
+                    return_value.value = {};
+                    return return_value;
+                }
 
-inline void set_value_multisz(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, const ::std::vector<::std::wstring>& data)
-{
-    const reg_view_details::reg_view regview{key};
-    return regview.set_value_multisz(subkey, value_name, data);
-}
+                // throw if exception policy
+                err_policy::HResult(HRESULT_FROM_WIN32(return_value.inner_error));
+                return return_value;
+            }
 
-inline void set_value_multisz(HKEY key, _In_opt_ PCWSTR value_name, const ::std::vector<::std::wstring>& data)
-{
-    return set_value_multisz(key, nullptr, value_name, data);
-}
+            template <typename R>
+            ::wil::reg::optional_value<R>& try_get_value(_In_opt_ PCWSTR value_name, DWORD type = ::wil::reg::reg_view_details::get_value_type<R>()) const
+            {
+                return try_get_value(nullptr, value_name, type);
+            }
 
-inline void set_value_multisz(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, const ::std::list<::std::wstring>& data)
-{
-    const reg_view_details::reg_view regview{key};
-    return regview.set_value_multisz(subkey, value_name, data);
-}
+            template <typename R>
+            typename err_policy::result set_value(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, const R& value) const
+            {
+                const auto error = ::RegSetKeyValueW(
+                    ::wil::reg::details::get_key(m_key),
+                    subkey,
+                    value_name,
+                    ::wil::reg::reg_view_details::set_value_type(value),
+                    static_cast<BYTE*>(::wil::reg::reg_view_details::get_buffer(value)),
+                    ::wil::reg::reg_view_details::get_buffer_size(value));
+                return err_policy::HResult(HRESULT_FROM_WIN32(error));
+            }
 
-inline void set_value_multisz(HKEY key, _In_opt_ PCWSTR value_name, const ::std::list<::std::wstring>& data)
-{
-    return set_value_multisz(key, nullptr, value_name, data);
-}
-#endif
+            template <typename R>
+            typename err_policy::result set_value(_In_opt_ PCWSTR value_name, const R& value) const
+            {
+                return set_value(nullptr, value_name, value);
+            }
 
-inline HRESULT set_value_dword_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, DWORD data) WI_NOEXCEPT
-{
-    const reg_view_details::reg_view_nothrow regview{key};
-    return regview.set_value(subkey, value_name, data);
-}
+#if defined(_VECTOR_) && defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+            template <typename R>
+            typename err_policy::result set_value_multistring(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, const R& data) const try

```

Comment:
Does this suggest that `set_value` calls won't work for multistring? Is there anyway we could provide this code as a template specialization of `set_value`?